File: poke-api.py
from flask import Flask
from flask_restful import request ,Resource, Api
import requests

POKE_API_URI_LIST = 'https://pokeapi.co/api/v2/pokemon/'

app = Flask(__name__)
api = Api(app)


class PokeList(Resource):
    def get(self):
        query_params = {}
        query_params['offset'] = request.args.get('offset')
        query_params['limit'] = request.args.get('limit')
        name_to_search = request.args.get('q')

        if len(query_params):
            target_uri = POKE_API_URI_LIST[:-1] + "?"
            first = True
            for param in query_params:
                if first:
                    first = False
                    target_uri = target_uri + param + "=" + query_params[param]
                else:
                    target_uri = target_uri + "&" + param + "=" +  query_params[param]                 
        else:
            target_uri = POKE_API_URI_LIST

        if name_to_search == None:
            poke_response = requests.get(target_uri).json()

            poke_list_result = []
            poke_list_received = poke_response['results']
            for pokemon in poke_list_received:
                pokemon_url = pokemon['url']
                pokemon_id = pokemon_url.split("pokemon/")[1][:-1]
                pokemon_name = pokemon['name']
                poke_data = requests.get(pokemon_url).json()
                poke_list_result.append(dict(
                    id = pokemon_id,
                    name = pokemon_name,
                    image = poke_data['sprites']['front_default']
                ))
            
            print(target_uri)
            final_response = dict(  total=poke_response['count'],
                                    limit=query_params['limit'],
                                    offset=query_params['offset'],
                                    data =  poke_list_result )
            return final_response



api.add_resource(PokeList, '/')

if __name__ == '__main__':
    app.run(debug=True)

File: poke-api-go/poke-api.go
package main

// Global imports
import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"
)

// Global constants
const pokeApiUrl = "https://pokeapi.co/api/v2/pokemon/"
const pokeSpriteUrl = "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/"

// Structs to receive data from the pokeapi
type pokeApiList struct {
	Count    int
	Next     string
	Previous string
	Results  []pokeData
}
type pokeData struct {
	Name string
	Url  string
}

// Structs to send data to clients
type pokeResult struct {
	Id    string `json:"id"`
	Name  string `json:"name"`
	Image string `json:"image"`
}
type fullPokeResult struct {
	Total  int          `json:"total"`
	Limit  int          `json:"limit"`
	Offset int          `json:"offset"`
	Data   []pokeResult `json:"data"`
}

// Main with the endpoint declarations
func main() {
	router := gin.Default()
	router.GET("/", getPokes)
	router.GET("/:name", getPokeByName)
	router.Run("localhost:8080")
}

// Home endpoint
func getPokes(c *gin.Context) {
	var listFinished bool = false
	var targetUrl string = pokeApiUrl
	var result fullPokeResult
	var countPokes int
	var pokeDraft pokeResult
	var apiResponse pokeApiList

	// Parsing query parameters
	limitStr := c.DefaultQuery("limit", "20")
	offsetStr := c.DefaultQuery("offset", "0")
	qName := c.DefaultQuery("q", "")

	// Converting limit and offset to integers
	limitInt, e1 := strconv.Atoi(limitStr)
	if e1 != nil {
		c.IndentedJSON(http.StatusBadRequest, gin.H{
			"message": "Invalid request"})
		return
	}
	offsetInt, e1 := strconv.Atoi(offsetStr)
	if e1 != nil {
		c.IndentedJSON(http.StatusBadRequest, gin.H{
			"message": "Invalid request"})
		return
	}

	// Count and limit for the response are already known
	result.Limit = limitInt
	result.Offset = offsetInt

	// Setting up target Url for fetching the list of pokes
	targetUrl = targetUrl + "?offset=" + offsetStr

	// Main loop to fetch the data
	for !listFinished {
		// Perform the http request and parse it
		fmt.Printf("poke-server: fetching from %s\n", targetUrl)
		res, err := http.Get(targetUrl)
		if err != nil {
			fmt.Printf("poke-server: error making http request: %s\n", err)
			os.Exit(1)
		}
		fmt.Printf("poke-server: got response!\n")
		fmt.Printf("poke-server: status code: %d\n", res.StatusCode)
		resBody, err := io.ReadAll(res.Body)
		if err != nil {
			fmt.Printf("poke-server: could not read response body: %s\n", err)
			os.Exit(1)
		}
		err = json.Unmarshal([]byte(resBody), &apiResponse)
		if err != nil {
			fmt.Printf("poke-server: could not parse json data: %s\n", err)
			os.Exit(1)
		}

		//Loop over the list of pokes we received filtering if needed
		for _, v := range apiResponse.Results {
			vSplit := strings.Split(v.Url, "/")
			id := vSplit[len(vSplit)-2]
			pokeDraft.Id = id
			pokeDraft.Image = pokeSpriteUrl + string(id) + ".png"
			pokeDraft.Name = v.Name
			if qName != "" {
				if !strings.Contains(v.Name, qName) {
					continue
				}
			}
			result.Data = append(result.Data, pokeDraft)
			countPokes++
			if countPokes == limitInt {
				break
			}
		}

		// Check if ending condition happened (no more pokes or reached limit)
		if apiResponse.Next == "" || countPokes == limitInt {
			listFinished = true
		} else {
			targetUrl = apiResponse.Next
		}
	}

	// Sending the response
	result.Total = apiResponse.Count
	fmt.Printf("poke-server: full pokemon list fetched from server\n")
	c.IndentedJSON(http.StatusOK, result)
}

func getPokeByName(c *gin.Context) {

}
